### ISSUE TITLE
Nut addon disconnects from device once a day

### ISSUE THREAD

**Issue by quarcko** [hassio-addons/addon-nut#109]
# Problem/Motivation

After some recent update NUT addon became unstable - once every day it
disconnects from device, and restart of addon is required. 
After restarting addon it works again for some time.
Here are logs, that just keep repeating:
141332.209845	Trying to connect to UPS [PowerWalker_VI_800@localhost]
141465.449589	UPS [PowerWalker_VI_800@localhost]: connect failed: Connection failure: Cannot assign requested address

## Expected behavior

Working stable as before, and not restarting every day

## Steps to reproduce

Just start the addon and it will fail after some time.

**Comment by sinclairpaul** [comment]
Please post the full addon startup log, and configuration.

**Comment by quarcko** [comment]
Here is config:

`users:
  - username: *
    password: *
    instcmds:
      - all
    actions: []
    upsmon: master
devices:
  - name: PowerWalker_VI_800
    driver: usbhid-ups
    port: auto
    config: []
mode: netserver
shutdown_host: 'true'`

and here startup log:

`[s6-init] making user provided files available at /var/run/s6/etc...exited 0.
[s6-init] ensuring user provided files have correct perms...exited 0.
[fix-attrs.d] applying ownership & permissions fixes...
[fix-attrs.d] nut: applying... 
[fix-attrs.d] nut: exited 0.
[fix-attrs.d] done.
[cont-init.d] executing container initialization scripts...
[cont-init.d] 00-banner.sh: executing... 
-----------------------------------------------------------
 Add-on: Network UPS Tools
 Manage battery backup (UPS) devices
-----------------------------------------------------------
 Add-on version: 0.6.2
 You are running the latest version of this add-on.
 System: Home Assistant OS 5.12  (armv7 / raspberrypi4)
 Home Assistant Core: 2021.3.4
 Home Assistant Supervisor: 2021.03.6
-----------------------------------------------------------
 Please, share the above information when looking for help
 or support in, e.g., GitHub, forums or the Discord chat.
-----------------------------------------------------------
[cont-init.d] 00-banner.sh: exited 0.
[cont-init.d] 01-log-level.sh: executing... 
[cont-init.d] 01-log-level.sh: exited 0.
[cont-init.d] nut.sh: executing... 
[09:45:22] INFO: Setting mode to netserver...
[09:45:22] INFO: Generating /etc/nut/upsd.users...
[09:45:22] INFO: Configuring user: ****************
[09:45:22] INFO: Password is NOT in the Have I Been Pwned database! Nice!
[09:45:23] INFO: Configuring Device named PowerWalker_VI_800...
[09:45:23] INFO: Starting the UPS drivers...
Using subdriver: CyberPower HID 0.4
Network UPS Tools - Generic HID driver 0.41 (2.7.4)
USB communication driver 0.33
Network UPS Tools - UPS driver controller 2.7.4
[09:45:23] WARNING: UPS Shutdown will shutdown the host
[cont-init.d] nut.sh: exited 0.
[cont-init.d] nutclient.sh: executing... 
[cont-init.d] nutclient.sh: exited 0.
[cont-init.d] done.
[services.d] starting services
[services.d] done.
[09:45:24] INFO: Starting the UPS information server...
   0.000000	fopen /var/run/nut/upsd.pid: No such file or directory
   0.001473	listening on 0.0.0.0 port 3493
   0.003509	Connected to UPS [PowerWalker_VI_800]: usbhid-ups-PowerWalker_VI_800
[09:45:24] INFO: Starting the UPS monitor and shutdown controller...
   0.000000	fopen /var/run/nut/upsmon.pid: No such file or directory
   0.000645	Using power down flag file /etc/killpower
   0.000987	UPS: PowerWalker_VI_800@localhost (master) (power value 1)
   0.001228	debug level is '1'
   0.001272	Warning: running as one big root process by request (upsmon -p)
   0.001891	Init SSL without certificate database
   0.006208	Trying to connect to UPS [PowerWalker_VI_800@localhost]
   0.032250	User upsmonmaster@127.0.0.1 logged into UPS [PowerWalker_VI_800]
   0.009270	Logged into UPS PowerWalker_VI_800@localhost`

**Comment by sinclairpaul** [comment]
Assuming the name of your UPs is the same model, looking at the NUT documentation they recommend a different driver.

https://networkupstools.org/stable-hcl.html

Unfortunately this is likely a specific hardware/config issue to which I cannot do anything about.  The change may have been the way the Supervisor now presents hardware, or the changes to the addon to support this and the related issues raised, but neither of which we can change.

**Comment by quarcko** [comment]
I think there is simply an error on their web page, as recommended driver does not work.
Secondly, this addon was working well with same config from start like 2 years ago when it was even unreleased.

Also, your addon consists of 2 executable processes:
INFO: Starting the UPS information server...
INFO: Starting the UPS monitor and shutdown controller...

And my error clearly states, that:
"monitor and shutdown controller" cannot connect to "information server".
and not that "information server" cannot connect to HW.

My log:
141332.209845 **Trying to connect to UPS [PowerWalker_VI_800@localhost]**
141465.449589 UPS [PowerWalker_VI_800@localhost]: connect failed: Connection failure: Cannot assign requested address

startup log part of "monitor and shutdown controller":
0.006208 **Trying to connect to UPS [PowerWalker_VI_800@localhost]**
0.032250 User upsmonmaster@127.0.0.1 logged into UPS [PowerWalker_VI_800]
0.009270 Logged into UPS PowerWalker_VI_800@localhost


So i assume that "information server" just dies somehow.
and then no information in HA and thus errors in your addon
as one part cannot communicate with other part.

And simple restart of addon makes it work again.
So clearly problem is somewhere in the addon i think.
Or you should add a watchdog, when process dies to restart it?

**Comment by quarcko** [comment]
A bit more info: 
while addon is in "failed" state, connecting to container via portainer
and executing this: `upsdrvctl -u root start`
makes addon work again without restarting.
This is what is printed after executing command:

`
root@a0d7b954-nut:/# upsdrvctl -u root start
Network UPS Tools - UPS driver controller 2.7.4
Network UPS Tools - Generic HID driver 0.41 (2.7.4)
USB communication driver 0.33
Duplicate driver instance detected! Terminating other driver!
Using subdriver: CyberPower HID 0.4
root@a0d7b954-nut:/# 
`

So yeah, you were right, driver is "loosing" connection to device, but driver process is still running
and gets restarted only after manual intervention...

will do an automation in HA for now, to monitor UPS state and just restart addon.

**Comment by sinclairpaul** [comment]
Maybe look at the polling interval settings for the device, it's interesting that it states it is using the Cyberpower sub driver.

You will find a few other issues relating to the polling and those UPS devices.

**Comment by apeeters** [comment]
For reference: the latest release [0.6.2](https://github.com/hassio-addons/addon-nut/releases/tag/v0.6.2) adds a new configuration option `upsmon_deadtime`.

Since I have the same issue with a CyberPower UPS, I tried setting this option to 25 as suggested elsewhere. This seems to have resolved the issue. 🤞

https://github.com/networkupstools/nut/issues/735
https://raspberrypi.stackexchange.com/questions/66611/nut-cyberpower-data-stale
https://nmaggioni.xyz/2017/03/14/NUT-CyberPower-UPS/

**Comment by apeeters** [comment]
Update: the issue was not resolved 😢

**Comment by sinclairpaul** [comment]
```
devices:
  - name: Cyberpower
    driver: usbhid-ups
    port: auto
    config:
      - pollinterval = 15
mode: netserver
shutdown_host: 'false'
upsd_maxage: 25
upsmon_deadtime: 25
```

I have run this config for months with no timeouts.  Rather than comment on old issues, if you still need assistance please open a new one 😉

**Comment by apeeters** [comment]
I also found these settings, even without `upsd_maxage` it is now running fine. Relevant parts of my config:

```
devices:
  - <snip>
    config:
      - pollinterval = 15
upsmon_deadtime: 25
```

@sinclairpaul I did not want to raise a new issue since the problem is not related to the addon but rather upstream NUT. I was commenting here to help future readers of this issue 😃
